[1~5단계 - 로또] 권장순 미션 제출합니다.

jsoonworld opened the pull request:
# 안녕하세요, 오찌~! 😊

이번 미션에서는 **단계별 요구사항을 충실히 지키는 것**을 가장 중요한 목표로 삼아 진행해보았어요!  
진행하면서도 "지금 구조가 과한 건 아닐까?", "너무 나누기만 한 건 아닐까?" 하는 고민이 많았고,  
그 고민의 흐름을 **Step별로 정리**해보았습니다.

혹시 보시다가 궁금한 부분이나 아쉬운 부분,  
혹은 더 나은 방향이 있다면 부담 없이 편하게 피드백 주시면 정말 감사하겠습니다 🙏 

---

# Step 1. 요구사항부터 제대로 지켜보자!

지난 대면 미팅에서 오찌가 해주셨던 말이 인상 깊었어요.

> "요구사항을 지키지 않고 하고 싶은 기술만 사용하다가, 결국 요구사항을 만족하지 못한다면 그동안은 무용지물이 되고, 결국 할 일을 못한 것이 된다."

이 말씀을 듣고 저는 이렇게 받아들였어요.  
**"요구사항을 제대로 지켜야, 그 다음이 있다!"**

그래서 이번 미션에서는 **요구사항을 최대한 충실히 지키는 것**을 가장 중요한 기준으로 삼아 구현을 진행해보았어요.

- 처음에는 단순히 숫자 생성 → 티켓 생성 흐름까지만 구현했지만,
  테스트 코드를 작성하면서 불편함을 느껴 `LottoNumberGenerator`, `LottoGenerator`, `LottoTicketGenerator` 등으로 **구조를 나누고 책임을 분리**하게 되었어요.
- 특히 숫자 정렬 책임을 어디에 둘지가 고민이었고,  
  이렇게 추상화를 한 구조가 **요구사항에 비해 과한 건 아닐까?** 하는 걱정도 들었습니다.

📌 **오찌께 여쭤보고 싶은 점**
- 현재 구조가 요구사항에 비해 과도하게 분리된 건 아닌지,  
- 혹시 더 간결하게 정리할 수 있는 부분이 있다면 피드백 부탁드릴게요!

---

# Step 2. 요구사항에 없는 기능도 구현해야 할까?

이번 단계에서는 **요구사항을 넘어서는 구현을 어디까지 할지**에 대한 고민이 많았어요.

요구사항에는 **입력 재시도나 유효성 검증, 테스트 코드 작성**에 대한 내용이 따로 명시되어 있지 않아서, 이 부분까지 구현하는 것이 맞는지 망설여졌습니다.

예를 들어 `"14000"`이라는 값이 음수이거나 숫자가 아닌 경우,  
이걸 누가 검증해야 할지 헷갈리더라고요.

- `view`에서 처리해야 할까?
- `parser`에서 가져가야 할까?
- 아니면 도메인 객체인 `generator` 쪽에서 책임져야 할까?

그리고 `LottoNumber`와 `WinningNumber`는 역할이 비슷해서  
**상속으로 중복을 줄이고 싶다는 생각도 들었지만**,  
상속에 대한 거부감도 있어서 결국 중복을 감수하고 별도로 분리했어요.

- 중복을 허용하고 각각의 의미를 살릴 것인지,  
- 아니면 역할이 유사하니 추상화를 고려할 것인지,  
  여전히 고민이 많이 되는 부분입니다.

📌 **오찌께 여쭤보고 싶은 점**
- 이런 경우, 입력값 검증 책임은 어느 계층에 두는 것이 가장 자연스럽다고 느끼시나요?
- 요구사항에 없는 **재입력 처리나 테스트 코드 작성, 유효성 검증**까지 구현하는 건  
  실무에서도 바람직한 접근인지, 아니면 요구사항을 우선시하는 게 맞을지 오찌의 생각이 궁금해요!
- 그리고 역할이 유사한 도메인 객체에 대해서 **상속을 활용해 중복을 줄이는 방식**과,  
  지금처럼 **의미를 분리해서 각각 유지하는 방식** 중 어느 쪽이 더 적절했을지도 여쭙고 싶습니다!

---

# Step 3. 테스트 코드의 커버리지와 구성

이번 미션에서는 **테스트 코드 구성에도 조금 더 신중하게 접근**해보았어요.

이전 미션에서는 테스트가 너무 많아 오히려 의미가 흐려졌던 경험이 있어서,  
이번에는 **정말 필요한 테스트 위주로 구성해보자**는 생각으로 접근했어요.

- 특히 **경계값 중심의 테스트**에 집중했고,  
  불필요한 중복 테스트는 최대한 줄이려고 했어요.
- 그리고 테스트 헬퍼 메서드도 여러 테스트에서 재사용할 수 있긴 했지만,  
  오히려 **공통화가 유지보수를 어렵게 만들 수 있다**는 우려가 들어서  
  각 테스트마다 **필요한 데이터만 명확히 선언**하는 방식을 택했어요.

그 결과 테스트의 **가독성은 올라간 것 같은데**,  
이게 정말 적절한 커버리지인지, 아니면 부족한 부분이 있는지는 스스로도 아직 확신이 없어요 😅

📌 **오찌께 여쭤보고 싶은 점**
- 현재 테스트 커버리지나 구성 방식이 적절한지 궁금합니다!
- 혹시 보시기에 **더 필요한 테스트나 불필요한 테스트**가 있다면 피드백 부탁드릴게요 🙏
- 테스트마다 개별 데이터를 선언한 방식이 **가독성과 유지보수 측면에서 괜찮았는지**,  
  아니면 오히려 공통화가 더 나았을지 오찌의 의견이 궁금합니다!

---

# Step 4~5. 수동 기능 추가와 구조 확장 고민

이번 단계에서는 **자동 구매 로직 위에 수동 구매 기능을 추가**하면서  
기존 구조와의 연결, 책임 분리 등에서 생각보다 많은 고민이 들었어요.

- 처음엔 Controller 내부를 메서드 단위로 나누어 깔끔하게 구성했다고 생각했는데,  
  수동 기능이 추가되면서 흐름이 **점점 복잡해지고 무거워졌고**,  
  결국 일부 로직을 **Service나 Util 쪽으로 분리**하게 되었습니다.
- 하지만 그렇게 나누다 보니 이제는 **Service와 Util이 복잡해진 건 아닐까?**  
  라는 고민이 또 생기더라구요.
- 역할과 책임에 따라 한 가지 기능만 수행하도록 나누었다고 생각했는데,  
  전체 구조를 펼쳐보면 **요구사항에 비해 지나치게 많은 추상화와 분리**가 이뤄진 것처럼 보이기도 했어요.

또 한 가지는 **정적 팩토리 메서드 사용에 대한 고민**이었어요.

- 그동안은 습관처럼 사용해왔는데, 이번에는 꼭 사용하지 않아도 되는 경우에는  
  **단순히 복잡도를 높이지 않도록 사용을 줄여보자**는 판단을 했어요.
- 특히 도메인, DTO에서도 정적 팩토리 메서드를 적용할 수 있겠다는 생각이 들었지만,  
  굳이 사용하지 않아도 의미 전달이 가능하다면 생략하는 쪽을 택했습니다.

추가로 계층 간 역할 분리도 아직 명확하지 않게 느껴졌어요.

- `view → util → dto → domain/service` 흐름 안에서  
  **포맷팅이나 출력 책임이 어디에 위치해야 할지** 헷갈리는 순간이 많았고,
- DTO에 포맷팅을 넣자니 계층 책임과 맞지 않고,  
  Util에 넣자니 너무 많은 역할을 주는 것 같아 **서비스 계층으로 올려야 할지** 고민이 들었어요.
- 최대한 담백하게 구현하려고 했지만, 역할과 책임을 나누다 보니  
  **오히려 결합도가 지나치게 낮아져**서  
  "다시 합치는 게 더 나은 건 아닐까?"라는 생각까지 들기도 했습니다 😅

📌 **오찌께 여쭤보고 싶은 점**
- 지금처럼 구조를 나누고 인터페이스를 도입한 방식이  
  **오히려 과한 추상화는 아닌지** 궁금합니다!
- 구조적으로 너무 쪼개져서 **요구사항에 비해 과하게 나눴다는 느낌이 드는데**,  
  혹시 오찌가 보시기에 **조금 더 단순화해도 괜찮은 부분**이 있다면 조언 부탁드릴게요.
- 그리고 정적 팩토리 메서드는 **어떤 기준으로 사용하는 것이 좋을지**,  
  오찌는 어떤 상황에서 적극적으로 사용하는지도 듣고 싶습니다!
- 마지막으로, `view`, `util`, `dto`, `domain/service` 계층 간의 **출력 책임 분리 기준**이  
  아직 불분명하게 느껴지는데, 오찌는 어떤 기준으로 나누고 계신지도 궁금해요 🙇‍♀️

---

# 마무리하며

이번 미션은 저에게 **가장 아쉬움이 많이 남는 미션**이었던 것 같아요.

처음엔 "요구사항을 최대한 담백하게 지키자!"는 마음으로 시작했는데,  
막상 구현을 마치고 나니 **가장 복잡하고, 일관성도 떨어지는 구조**가 되어버린 것 같다는 생각이 들어요.

오히려 지금 이 상태에서 리팩토링을 시도하는 것보다,  
**처음부터 다시 구현하는 것이 더 나은 구조를 만들 수 있지 않을까?**  
라는 생각이 들 정도로 아쉬움이 컸습니다 😢

그럼에도 불구하고, 그 과정에서 **객체지향에 대해 더 깊이 고민하고**,  
"역할과 책임을 어떻게 나눌 것인가", "요구사항을 어떻게 해석하고 구현할 것인가"에 대해  
스스로 많은 질문을 던지고 정리해볼 수 있었던 시간이었다고 생각해요.

오찌의 시선으로 보았을 때 어떤 부분이 아쉬웠는지,  
혹은 지금의 고민들이 좋은 방향으로 나아가고 있는 건지 피드백 주신다면  
앞으로 더 나아지는 데 큰 힘이 될 것 같습니다! 🙇‍♀️

긴 글 읽어주셔서 감사합니다 😊

jsoonworld committed:
feat(Lotto) : 1단계 - 로또 자동 구매

jsoonworld committed:
feat(Lotto) : 2단계 - 로또 당첨

jsoonworld committed:
feat(Lotto) : 3단계 - 로또 2등 당첨

jsoonworld committed:
feat(Lotto) : 4단계 - 로또 수동 구매

jsoonworld committed:
refactor(Lotto) : 5단계 - 리팩터링

Ohzzi commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,16 @@
+package domain;
+
+public class Lotto {
+    public static final int PRICE = 1000;
+
+    private final LottoNumbers numbers;
+
+    public Lotto(LottoNumbers numbers) {
+        this.numbers = numbers;
+    }
+
+    public LottoNumbers getNumbers() {
+        return numbers;
+    }
+}
+
```
로또는 LottoNumbers의 프록시일 뿐인 것 같은데요. LottoNumbers가 존재할 필요가 있을까요? `6개의 중복되지 않은 로또 번호를 가진다`라는 규칙은 로또가 가지고 있어도 충분히 합리적일 것 같아요

jsoonworld replied:
피드백 감사합니다! 😊  

말씀 듣고 보니 Lotto가 단순히 LottoNumbers를 감싸기만 하고 있어서 굳이 객체를 분리하지 않아도 될 것 같다는 생각이 들었어요. 처음엔 책임을 나눠야 한다고 생각했는데, 로또 번호의 유효성이나 중복 검증 같은 규칙을 Lotto가 직접 갖고 있어도 충분히 자연스럽고 명확할 것 같아요. 구조도 더 단순해지고 역할도 더 또렷해질 것 같아서, LottoNumbers를 제거하고 로직을 Lotto 안으로 옮기는 방향으로 리팩터링해보려고 합니다.

인사이트 감사합니다! 🙏

Ohzzi commented on src/main/java/domain/WinningNumber.java:
```diff
@@ -0,0 +1,36 @@
+package domain;
+
+public class WinningNumber {
```
당첨 번호는 로또 번호가 아닐까요? WinningNumber가 의미적으로나 기능적으로나 LottoNumber와 가지는 차이점이 애매한 것 같아요.

jsoonworld replied:
피드백 감사합니다! 😊  

말씀해주신 내용을 읽고 나서, 저도 WinningNumber와 LottoNumber의 경계가 확실하지 않다는 점에 깊이 공감했어요.  
사실 처음에는 구조를 나누는 게 구현상 편할 것 같아서 별도로 분리했는데, 지금 돌아보면 책임과 역할에 대한 고민이 충분하지 않았던 것 같아요. 말씀 덕분에 의미적으로 하나의 개념으로 볼 수 있다는 점을 다시 생각해보게 되었고, 더 적절한 형태로 개선해보겠습니다!  

꼼꼼한 피드백 정말 감사드립니다 🙏

Ohzzi commented on src/main/java/domain/Profit.java:
```diff
@@ -0,0 +1,47 @@
+        return total;
+    }
+
+    private List<Rank> winningRanks() {
+        return List.of(Rank.FIFTH, Rank.FOURTH, Rank.THIRD, Rank.SECOND, Rank.FIRST);
+    }
```
이친구는 상수화 되어 있는 것이 맞지 않을까요? calculateTotalPrize 할 때마다 리스트를 새로 만들 이유가 없을 것 같아요.

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, `winningRanks()` 메서드에서 리스트를 매번 새로 생성할 필요 없이 상수로 정의하는 것이 더 효율적이고, 코드의 성능도 개선될 수 있을 것 같아요. 리스트를 매번 새로 만드는 이유는 성능에 비효율적이고, 상수화된 값이 적절하게 사용될 수 있겠다는 생각이 들었습니다.

`winningRanks`를 상수화하여 재사용할 수 있도록 리팩토링하는 방향으로 개선해보겠습니다. 

꼼꼼한 피드백 감사합니다!

Ohzzi commented on src/main/java/domain/BonusNumber.java:
```diff
@@ -0,0 +1,47 @@
+package domain;
+
+public class BonusNumber {
```
이 클래스도, 보너스 넘버는 결국 로또 번호인데 왜 클래스가 분리되어 있을까요? 또한 `보너스 번호가 당첨 번호와 중복될 수 없다` 라는 요구사항을 위해 보너스 번호 생성자가 WinningNumbers까지 받고 있어요. 생성자가 해야 할 역할이 맞을까요? 보너스 번호 클래스가 과연 정말 분리가 필요한지 고민해볼 필요가 있을 것 같아요

jsoonworld replied:
피드백 감사합니다! 😊

저는 처음에는 로또 번호 6개와 보너스 번호 1개가 다르다고 생각해서 분리하는 것이 맞다고 판단했어요. 로또 번호와 보너스 번호가 역할적으로 다르다고 느껴서 각기 다른 클래스로 분리해야겠다고 생각했죠.

하지만 말씀하신 것처럼, `BonusNumber` 클래스가 보너스 번호의 생성자에서 당첨 번호와의 중복을 검증하는 역할을 맡고 있다는 부분은 조금 과도할 수 있겠다는 생각이 듭니다. 또한, 보너스 번호가 사실 로또 번호의 특수한 형태라는 점에서, 굳이 분리할 필요가 없을 수도 있다는 점을 고민하게 되었어요.

따라서, 이 부분을 리팩터링하여 `BonusNumber` 클래스를 불필요하게 분리하지 않도록 개선해보겠습니다.  
좋은 고민을 던져주셔서 감사합니다! 🙏

Ohzzi commented on src/main/java/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.List;
+import java.util.Set;
+
+public class WinningNumbers {
+    private static final int REQUIRED_SIZE = 6;
+    private static final String ERROR_INVALID_SIZE = "[ERROR] 당첨 번호는 6개여야 합니다.";
+    private static final String ERROR_DUPLICATED = "[ERROR] 당첨 번호는 중복될 수 없습니다.";
+
+    private final List<WinningNumber> numbers;
```
당첨 번호가 보너스 번호까지 알고 있는게 요구사항을 검증하기 더 좋지 않을까요?

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, 당첨 번호가 보너스 번호까지 알고 있다면 요구사항을 검증하는 데 더 유리할 것 같다는 생각이 드네요.  
검증이 필요하다면 무조건 분리하는 것보다는 누가 어떤 검증을 책임질 것인지까지 고민해봐야 한다는 점을 다시 한번 깨닫게 되었습니다! 감사합니다!

이전에 주신 피드백을 바탕으로, `WinningNumber` 대신 `LottoNumber`로 당첨 번호를 관리하는 방식으로 리팩터링을 진행했고, 조금 더 검증이 용이한 방향으로 개선된 것 같습니다.

좋은 피드백 덕분에 설계를 더욱 개선할 수 있었습니다. 감사합니다! 🙏

Ohzzi commented on src/main/java/service/assembler/ManualLottoAssembler.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ManualLottoAssembler {
+
+    public static List<Lotto> assemble(int manualCount, InputView inputView, OutputView outputView) {
```
assembler, assemble이라는 네이밍은 어떠한 의도에 의해 정해진 것인가요?

jsoonworld replied:
질문 감사합니다!

수동으로 로또를 뽑아야 한다는 요구사항이 추가되면서 기능을 구현하려고 했을 때, 이미 코드에는 수동으로 입력을 받으면 이를 처리해줄 객체들이 존재하는 상황이었습니다. 그래서 수동으로 구매할 티켓의 갯수와 번호들을 적절히 배치하여 조합하는 역할만 남았다고 판단하게 되었어요.

이 역할을 수행하는 클래스와 메서드를 `어셈블러(Assembler)`와 `어셈블(Assemble)`이라는 네이밍으로 작성하게 되었고, 이 이름이 해당 기능을 잘 나타낼 수 있다고 생각했습니다!

Ohzzi commented on src/main/java/service/LottoPurchaseService.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface LottoPurchaseService {
+    Lottos purchase(LottoPurchaseDto request);
+}
```
구현체를 살펴보니까 LottoPurchaseServiceImpl밖에 없는 것 같아요. 이친구는 왜 구체 클래스가 아니라 인터페이스로 된 것일까요?

jsoonworld replied:
피드백 감사합니다! 😊

사실 서비스 코드에서는 연관 관계를 느슨하게 유지하는 것이 좋다고 생각해왔습니다. 서비스 로직에는 다양한 역할이 존재하고, 서비스 자체가 여러 가지를 연결해주는 역할을 하기 때문에, 한 가지 변경이 있을 때 여러 코드에 영향을 미치는 것을 피하고 싶었습니다. 이로 인해 `LottoPurchaseService`는 인터페이스로 추상화한 것입니다.

물론 요구사항만 본다면 과하게 느껴질 수 있다는 점은 이해하고 있습니다. 그런데 서비스 구현을 할 때면 항상 손이 먼저 인터페이스를 작성하고 있는 저를 발견하게 되네요. 😭

Ohzzi replied:
추상화는 적절히 사용하면 강결합을 방지하고 확장성에 도움이 된다는 장점이 있지만, 반대로 한 번에 로직을 이해하기 힘들다는 단점도 있어요. 인터페이스를 사용할 것인지 사용하지 않을 것인지는 개인의 취향이고 선택이지만, 개인적으로는 과한 추상화는 오히려 독이 된다고 생각해요! 가독성을 어느 정도 포기하면서 확장성을 챙기는 것인데, 확장성을 챙길 필요가 없는 경우에 가독성을 포기할 이유는 없으니까요 😄 `우선은 구체 클래스로 만든다` 라는 원칙을 세워보면 도움이 될거에요

jsoonworld replied:
말씀해주신 이 부분이 특히 와닿았습니다.

> "확장성을 챙길 필요가 없는 경우에 가독성을 포기할 이유는 없으니까요 😄"

불필요한 추상화는 오히려 이해를 어렵게 만들 수 있다는 점에 공감했고, 저도 현재는 '구체 클래스로 먼저 구현하고, 필요할 때 추상화한다'는 원칙을 가져가는 게 좋겠다고 느꼈어요.

지금은 구조를 근거 있게 나누고, 실제로 필요한 경우에만 확장을 시도하는 연습이 필요한 시기인 것 같아요. 말씀 주신 방향대로 차근차근 정리해보겠습니다! 😊

Ohzzi commented on src/main/java/service/TicketGenerator.java:
```diff
@@ -0,0 +1,5 @@
+package service;
+
+public interface TicketGenerator {
+    int generate(int amount);
+}
```
"Ticket"Generator 인데 generate 하는 값은 integer네요?

Ohzzi replied:
또한 파라미터 이름이 좀 더 구체적이면 어떨까요?

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, `TicketGenerator`의 반환 타입이 `int`인 점이 직관적이지 않다는 부분에 공감합니다. `TicketGenerator`가 티켓을 생성하는 역할이라면, 반환 타입을 `int`보다는 티켓 객체나 티켓 리스트로 변경하는 것이 더 명확할 것 같아요.

또한, 파라미터 이름 `amount`가 구체적이지 않다는 점도 이해가 됩니다. 이를 `purchaseAmount`와 같이 더 명확하게 구체화하는 것이 가독성에 도움이 될 것 같습니다.

따라서, `generateTickets` 메서드로 티켓 리스트를 반환하도록 개선하고, 파라미터 이름도 `purchaseAmount`로 변경하여 코드를 리팩터링 해보겠습니다!

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, `TicketGenerator`에서 반환값이 `int`인 점과 파라미터 이름이 구체적이지 않다는 부분에 대해 다시 한 번 고민하게 되었습니다. 

`Ticket`을 생성하는 기능은 필요하지만 객체를 추가하는 것이 부담스러워서, 이 부분을 어떻게 개선할지 고민 중입니다. 로또 생성과 관련된 책임을 `LottoGenerator`로 이전하는 것이 더 적절할 것 같다는 생각도 듭니다.

피드백을 통해 반환값과 클래스 간의 관계, 파라미터 네이밍을 좀 더 구체화해야 한다는 점에 대해 공감이 갔습니다. 이 부분을 개선할 방법을 생각하면서 리팩터링을 진행해보겠습니다.

다시 한 번 피드백 주셔서 감사합니다! 🙏

Ohzzi commented on src/main/java/service/OutputPresenter.java:
```diff
@@ -0,0 +1,11 @@
+import domain.Lottos;
+import domain.Profit;
+
+public interface OutputPresenter {
+    void showPurchasedLottos(int manualCount, Lottos allLottos);
+    void showStatistics(LottoStatistics statistics, Profit profit);
+}
```
뷰를 나타내는 것은 OutputView의 책임인가요 OutputPresenter의 책임인가요?

jsoonworld replied:
피드백 감사합니다! 😊

말씀 주신 내용을 보고 다시 생각해보게 되었어요. 처음엔 도메인 객체를 직접 `OutputView`에 넘기는 게 어색하다고 느껴서, 중간에서 가공해주는 역할로 `OutputPresenter`를 두었는데요, 지금 돌아보면 출력 책임은 `View` 쪽에서 전담하는 게 더 자연스러울 수 있겠다는 생각이 들었습니다.

출력 책임을 `OutputView`로 다시 옮기는 방향으로 한번 개선해보겠습니다! 🙏

Ohzzi commented on src/main/java/service/io/InputHandlerImpl.java:
```diff
@@ -0,0 +1,57 @@
+        return ManualLottoCountParser.parse(inputView.readManualLottoCount(), purchaseAmount);
+    }
+
+    @Override
+    public Lottos readManualLottos(int count) {
+        List<Lotto> manualLottos = ManualLottoAssembler.assemble(count, inputView, outputView);
+        return new Lottos(manualLottos);
+    }
+
+    @Override
+    public WinningNumbers readWinningNumbers() {
+        outputView.printLastWeekWinningNumbersPrompt();
+        return WinningNumbersParser.parse(inputView.readLastWeekWinningNumbers());
+    }
```
도메인 객체에 해당하는 Lottos와 WinningNumbers를 뷰의 일부인 `InputHandler`가 생성하는 것은 올바른 책임이라 볼 수 있을까요? 장순님이 생각하시는 뷰의 역할과 책임은 무엇인가요?

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, `InputHandler`가 도메인 객체인 `Lottos`와 `WinningNumbers`를 생성하는 것이 적절한지에 대해 다시 한 번 고민해보게 되었습니다.

저는 뷰(View)는 사용자와의 상호작용만 담당해야 한다고 생각합니다. 데이터 생성이나 비즈니스 로직 처리와 같은 책임은 **뷰**가 아닌 서비스나 도메인 계층에서 맡는 것이 맞다고 봅니다.

따라서, `InputHandler`가 도메인 객체를 직접 생성하는 대신, 사용자 입력을 받아 그 입력을 적절한 서비스나 도메인 로직에 위임하는 방향으로 개선할 필요가 있을 것 같습니다. 이 부분에 대해서는 좀 더 고민하고 개선해보겠습니다! 😊

다만, 이처럼 구현하려고 했던 의도는 **`view`가 출력만 해야 하는 역할**을 맡아야 한다는 점에서, 구체적인 반복문 로직 등을 포함하는 것이 부자연스럽다고 느꼈기 때문입니다. 그래서 점차적으로 **`view`의 역할을 도메인에 가까운 부분으로 이동시켜왔던 것 같습니다**.

Ohzzi commented on src/main/java/service/io/OutputPresenterImpl.java:
```diff
@@ -0,0 +1,48 @@
+    @Override
+    public void showPurchasedLottos(int manualCount, Lottos lottos) {
+        int autoCount = lottos.count() - manualCount;
+        LottoPurchaseSummaryDto summaryDto = new LottoPurchaseSummaryDto(manualCount, autoCount);
+
+        System.out.println();
+        outputView.printLottoPurchaseResultHeader(summaryDto.manualCount(), summaryDto.autoCount());
```
LottoPurchaseSummaryDto는 왜 생성되는 객체인가요? 지금 코드 구조에서는 전혀 필요하지 않은 객체로 보여요. 기껏 생성을 해놓고 바로 getter로 꺼내서 OutputView에 던져주고 있거든요. 사용처를 보건대 애초에 필요하지 않은 클래스였던 것은 아닐까요?

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, `LottoPurchaseSummaryDto` 객체를 생성한 후 바로 getter로 값을 꺼내서 `OutputView`에 넘기는 부분은 불필요하게 객체를 생성한 것 같다는 생각이 듭니다. 사실 이 부분은 **`OutputView`가 단순히 출력만 해야 한다**는 의도에서 복잡한 로직을 `view`가 아닌 다른 계층으로 이동시키려 했던 것 같습니다.

하지만, 이 객체가 불필요하게 생성되는 부분은 개선이 필요하다고 생각됩니다. 그래서 객체 생성 없이 직접 출력할 수 있도록 리팩토링을 진행해보겠습니다!

Ohzzi commented on src/main/java/service/generator/LottoGeneratorImpl.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class LottoGeneratorImpl implements LottoGenerator {
```
이름은 LottoGeneratorImpl인데 자동 로또 생성만 하네요

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, `LottoGeneratorImpl`라는 이름을 가지고 있지만 실제로 자동 로또만 생성하는 부분이 어색하다는 점에 공감합니다. 수동 로또를 추가하는 과정에서 기존 클래스들의 역할과 의미를 충분히 고려하지 못한 것 같네요 ㅜㅜ

피드백을 반영하여 클래스 이름과 책임을 더 명확하게 개선해보겠습니다! 감사합니다! 🙏

Ohzzi commented on src/main/java/dto/PrintableMatchDto.java:
```diff
@@ -0,0 +1,3 @@
+package dto;
+
+public record PrintableMatchDto(String messageLine) {}
```
단순 messageLine만 가지는 DTO의 필요 목적이 무엇인가요?

jsoonworld replied:
피드백 감사합니다! 😊

말씀해주신 대로, `PrintableMatchDto`와 같은 단순한 DTO가 필요한 목적에 대해 다시 한 번 고민하게 되었습니다. 제 의도는 뷰(View)에서 로직보다는 단순히 출력만 하도록 하기 위해, 전처리한 데이터를 DTO로 감싸서 넘겨주고자 했습니다.

하지만 이 부분이 사실 뷰의 역할에 더 가까운 것 같다는 생각이 들었습니다. 그래서 이 DTO가 정말 필요한지, 아니면 다른 방식으로 데이터를 넘기는 게 더 적절한지 리팩터링을 통해 개선해보겠습니다!

Ohzzi reviewed (changes requested):
장순님 미션 완료하느라 수고 많으셨습니다! 코드에 리뷰를 남기다가 중복되는 리뷰들이 많을 것 같아서 오늘은 본문 코멘트로 넘어와서 리뷰를 드릴게요. 질문사항에 대한 답변보다 우선 전체적인 코드에 대한 리뷰부터 남기도록 하겠습니다.

## 전체적인 코드에 대한 피드백

가감없이 피드백드리자면, 지난 미션 대비 이번 미션으로 넘어와서의 장순님의 코드가 훨씬 복잡해지고 읽기 어려워졌다고 느꼈습니다. 일단 이 미션의  제가 생각하기에 가장 큰 이유는 너무 많은 클래스입니다. 미션의 요구사항과 기능 대비 너무 많은 클래스가 분리된 것 같고, 명확한 역할을 가지지 않고 있는 클래스가 많습니다. 많은 클래스가 결코 좋은 것이 아닙니다. 클래스가 많아진다는 것은 그만큼 코드 흐름을 파악하기 위해 여러 클래스를 건너다녀야 한다는 것이고, 그 말인 즉 적절하지 못한 클래스의 분리는 가독성의 저해를 가져오게 됩니다.

전체적으로 보자면 도메인에서도 필요 없는 분리(Lotto - LottoNumbers 등)가 보입니다만, 가장 문제가 되는 부분은 제 기능을 한다고 보기 어려운 DTO와 각종 xxx`er` 클래스들의 남발이라고 생각합니다. 플로우 하나를 예로 들어보겠습니다. 로또 통계 하나를 보여주기 위해 다음의 과정을 거치고 있어요.
- LottoController -> OutputPresenter.showStatistics -> ResultViewModelAssembler.toPrintableMatchDtos, ResultViewModelAssembler.toPrintableProfitDto -> OutputView 호출하여 출력

이 과정이 정말 이렇게 많은 프레젠터, 어셈블러, 심지어 어셈블러 안에서는 Mapper까지 수많은 클래스가 필요한 것인지 다시 생각해보셨으면 좋겠습니다.

## 도메인과 뷰의 역할

도메인의 역할과 뷰의 역할이 뒤죽박죽 섞여있다고 생각합니다. 우선 짚고 넘어가야 할 것은, service 패키지 안에 있는 클래스들은 도메인의 역할을 하는 것인가요 뷰의 역할을 하는 것인가요? 도메인의 역할을 하는거라면 InputHandler와 OutputPresenter는 적절한 위치가 아닌 것 같습니다.

MVC 구조로 개발하는 것이라면, 도메인의 역할과 뷰의 역할은 서로 섞이면 안됩니다. 도메인의 역할과 책임은 도메인에 있고, 뷰의 역할과 책임은 뷰에 있어야 합니다. 그리고 그 사이를 컨트롤러 계층이 적절한 호출을 통해 엮어주는 역할을 할 뿐입니다. 하지만 지금의 코드는 도메인과 뷰가 상호 침범하는 케이스가 보이는 것 같아요.

## interface - implementation 구조
인터페이스를 분리한 추상화를 많이 사용해주셨는데요, 왜 인터페이스가 분리되었는지 명확하지 않은 경우가 보입니다. 인터페이스 분리는 의존성을 역전해야 하거나(지금 배울만한 내용은 아닙니다만), 다형성을 활용해서 다양한 구현체를 갈아끼울 필요가 있다거나, 테스트를 하기 용이하게 할 필요가 있다던가 하는 등 명확한 이유가 있어야 한다고 생각합니다. 그렇지 못하다면 그저 클래스 하나가 추가로 늘어나는 것 밖에 되지 않을 수도 있어요. 인터페이스를 분리해야 하는지 구체 클래스로 사용해야 하는지 정확한 기준을 세우고 그를 지켜가면서 개발하는 연습을 해보시면 좋을 것 같습니다.

질문해주셨던 내용에 대해서도 답변을 드려보면,

## 현재 구조가 요구사항에 비해 과도하게 분리된 건 아닌지, 혹시 더 간결하게 정리할 수 있는 부분이 있는지
이 부분은 바로 위의 피드백들로 답변을 대신할 수 있을 것 같습니다.

## 입력값 검증 책임 계층
`입력값을 검증하는 것`은 당연히 입력하는 계층에서 하는 것이 맞습니다. 그런데, 여기서 말하는 `입력값을 검증하는 것`은 입력이 적절한지이지 그것이 도메인 규칙을 위배하는지가 아닙니다. 예를 들어, 로또 번호는 1~45임을 검증하는 것이 표현 계층에 있는 것은 적절하지 않거나, 혹은 과한 validation이라고 생각합니다. 대신 로또 번호를 입력해야 하는데 문자열을 입력했다, 이러면 그건 input에서 검증해야 할 부분이겠죠. 이건 input으로 들어오는 것은 뭐 문자열 리터럴로 들어오든, 바이트 스트림으로 들어오든 도메인 안에서는 숫자로 존재하면서 1~45임을 검증해야 하는 것이고, 우리가 만든 프로그램의 구조상 문자열로 들어오기 때문에 숫자 문자열인지를 검증하는 것이 도메인 바깥에서 이뤄어져야 하는 것이라고 보면 됩니다.

## 요구사항에 없는 재입력 처리나 테스트 코드 작성, 유효성 검증까지 구현하는 건
실무에서도 바람직한 접근인지

당연하게도, `명시적으로 써있는 요구사항` 뿐 아니라 `당연한 규칙`도 검증하는 것이 맞습니다. 예를 들어 포인트 적립 시스템을 만드는데 기획서에 `포인트는 0원 미만일 수 없다`라고 적혀있지 않다고 0이상 검증을 안하는 것이 올바를까 라고 생각해보면 될 것 같습니다.

## 역할이 유사한 도메인 객체에 대해서 상속을 활용해 중복을 줄이는 방식과,
지금처럼 의미를 분리해서 각각 유지하는 방식 중 어느 쪽이 더 적절했을지

상속은 `유사한 클래스들 사이`에 적용하는 것이 아닙니다. 상속의 적절한 용례는 명확하게 부모 자식 관계일 때 적용하는 것이고 그마저도 최근에는 상속을 잘 사용하지 않는 것이 트렌드라면 트렌드라고 볼 수도 있을 것 같습니다. 보통은 상속보다는 조합(Composition)을 통해 작은 단위의 클래스들을 조합해서 의미를 만들어내곤 하죠.

다만, 현재 코드에서 `의미를 분리해서 각각 유지하는 방식`은 적절하지 못한 부분이 있습니다. 분리해야 하는 의미는 본질적인 측면의 의미라고 생각해요. 예를 들어 `LottoNumber`와 `WinningNumber`는 그 본질이 로또 번호임이 같고, 검증해야 하는 것이나 할 수 있는 역할도 같습니다. Lotto와 Winning 사이에 구분이 있어야 한다면 그건 `Lotto`와 `WinningLotto` 정도로 볼 수 있을 것 같네요. 이들은 로또 번호 리스트를 멤버로 가지는 것은 같지만 하는 역할이 다르니까요.

## 정적 팩토리 메서드는 어떤 기준으로 사용하는 것이 좋을지
정적 팩토리 메서드는 생성자에 명확한 네이밍을 부여할 필요가 있거나, 생성자와는 다른 방법으로 생성하고 싶을 때 사용하는 편입니다. 예를 들어, Lotto -> LottoDto 라고 변경을 할 일이 있다고 하면 LottoDto의 생성자를 그대로 사용하고 필드값들을 Lotto에서 getter로 꺼내줘서 넘기게 될텐데요. 이 로직을 LottoDto 안에 from 이라는 정적 팩토리 메서드로 만들어 외부로 노출하지 않고 사용하는 형태를 예시로 들어볼 수 있을 것 같아요.

## view, util, dto, domain/service 계층 간의 출력 책임 분리 기준이
아직 불분명하게 느껴지는데, 어떤 기준으로 나누면 좋을지
출력 책임은 view에 있는 것이 맞다고 생각합니다. 도메인 및 서비스에는 출력 책임이 전달되지 않아야 하는 것이 당연하다고 볼 수 있고요. 지금 책임이 헷갈리는 것은 자꾸 뷰의 영역을 도메인과 혼재해서 생각해서 그런거라고 생각합니다. 계속 말씀드리지만, 프로그램에 핵심적인 규칙이 아니라면 도메인에 넣지 않는다고 생각하면 맞습니다. 출력 포맷에 사소한 변경이 생긴다고 가정했을 때 어디까지 수정이 되어야 좋은 프로그램인가를 생각해보면 좋을 것 같습니다.

마지막으로 테스트에 대한 리뷰는 현재 상황에서는 생략하겠습니다. 프로덕션 코드에 대한 리뷰만 진행하고 해당 부분에 대한 장순님의 고민과 리팩토링이 모두 이어진 이후에 테스트 코드에 대한 리뷰를 진행해보도록 할게요!

jsoonworld committed:
refactor(Lotto) : LottoNumbers 제거 및 Lotto에 번호 검증 책임 통합

LottoNumbers 클래스를 제거하고 Lotto가 직접 List<LottoNumber>를 관리하도록 구조를 변경했습니다.
이에 따라 번호 개수 및 중복 검증 책임을 Lotto로 이전하고, 관련 생성자 및 검증 로직을 통합하였습니다.

- LottoNumbers → Lotto로 책임 이동
- Lotto 생성자에서 유효성(6개, 중복) 검증 수행
- 관련 서비스, 파서, 출력, 통계, 테스트 코드 전체 수정

jsoonworld committed:
refactor(WinningLotto) : `WinningNumbers`를 `WinningLotto`로 변경하고 관련 로직 리팩토링

- `WinningNumbers`를 `WinningLotto`로 대체하여 도메인 및 서비스 계층 리팩토링
- `BonusNumber`, `LottoStatistics`, 관련 테스트들을 `WinningLotto`에 맞게 수정
- `LottoStatisticsTest`, `ProfitTest`, `BonusNumberTest` 등에서 `WinningNumbers` 대신 `WinningLotto` 사용
- 파서 및 입력 처리 로직에서 `WinningLotto`를 사용하도록 변경
- `LottoNumber`와 `WinningLotto`의 일관성을 맞추기 위한 리팩토링

jsoonworld committed:
refactor(Profit) : winningRanks 메서드를 상수화하여 리스트 재사용

- `winningRanks()` 메서드를 상수 `WINNING_RANKS`로 대체하여 매번 리스트를 새로 생성하지 않도록 개선.
- 코드 성능을 개선하고, 상수화된 값을 재사용하여 불필요한 객체 생성을 방지.

jsoonworld committed:
refactor(LottoNumber) : BonusNumber를 LottoNumber로 변경

jsoonworld committed:
refactor(LottoGenerator) : TicketGenerator의 기능을 LottoGenerator로 이전

jsoonworld committed:
refactor(Presenter) : OutputPresenter, InputHandler 역할을 view로 이전

jsoonworld committed:
refactor(LottoGenerator) : 자동, 수동을 LottoGenerator에서 생성하도록 개선

jsoonworld committed:
refactor(result) : view의 책임을 갖고 있는 객체들을 개선하고 단순한 dto들을 제거

jsoonworld committed:
refactor(controller) : controller에 있던 비즈니스 책임을 서비스 클래스로 이전

jsoonworld committed:
refactor(test) : 리팩터링된 코드를 기반으로 테스트 코드 작성 및 개선

jsoonworld committed:
refactor(LottoService) : 메서드명을 간결하고 직관적으로 개선

Ohzzi commented on src/test/java/view/FakeInputView.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.List;
+
+public class FakeInputView implements InputView {
```
테스트 더블의 명명 규칙에서 Fake는 실제와 유사한 동작을 하도록 가짜 객체를 만드는 것인데요. 이 클래스는 실제와 유사한 동작을 한다기보다는 이미 지정된 응답을 반환하고 있는 것 같아요. 더 적절한 네이밍을 찾아볼까요?

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
말씀 듣고 보니, 현재 이 클래스는 실제 동작을 흉내내기보다는 정해진 값을 반환하는 역할에 더 가까운 것 같아요.

그래서 FakeInputView보다는 StubInputView처럼 고정된 응답을 제공하는 Stub이라는 의미가 더 적절하다고 판단했습니다.
이번 기회에 용어도 다시 한번 정리해봤어요 😊

- Dummy
테스트에서 사용되지 않지만, 메서드 호출을 위해 인자로 전달되어야 할 때 사용하는 객체.
예: 사용되지 않는 User 객체를 그냥 넘겨주는 경우

- Stub
미리 정의된 값을 반환하는 객체. 로직 없이 입력에 따라 고정된 출력만 제공.
예: 입력 값에 상관없이 "10000"을 반환하는 StubInputView

- Fake
실제 동작과 유사하지만, 간단하게 구현된 객체. 주로 인메모리 저장소나 간이 데이터베이스 형태로 사용됨.
예: 실제 DB 대신 사용하는 FakeRepository

- Mock
행위 기반 테스트에서 사용되며, 메서드 호출 여부나 호출 횟수를 검증하기 위해 사용됨.
예: send()가 몇 번 호출되었는지를 검증하는 MockEmailSender

이번 정리를 통해 저 자신도 더 명확하게 개념을 다질 수 있었고, 앞으로 더 적절한 네이밍과 구조를 고민하는 데 큰 도움이 될 것 같아요!

다시 한 번 감사드립니다 😊

Ohzzi commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public int countMatch(Lotto lotto) {
+        Set<Integer> winningValues = new HashSet<>();
+        for (LottoNumber number : numbers) {
+            winningValues.add(number.value());
+        }
+
+        long count = lotto.getNumbers().stream()
+                .map(LottoNumber::value)
+                .filter(winningValues::contains)
+                .count();
```
너무 어렵게 돌아간다는 느낌이 들어요! 그냥 lotto.getNumbers를 순회하면서 일치하면 count를 1씩 올려주는 방법을 쓰거나, 아니면 스트림 필터를 적절히 이용하면 될 것 같아요!

그리고, 지금처럼 LottoNumber의 value도 꺼내고 numbers의 value도 꺼내는건 VO를 적절하게 활용하지 못하는 것이 아닐까 하는 생각이 드네요!

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
말씀해주신 부분에 공감이 많이 갔어요.

지금 로직은 LottoNumber의 value()를 꺼내서 비교하고 있어서, VO의 의미가 제대로 드러나지 않고 단순 정수 비교처럼 보이는 문제가 분명 있었던 것 같아요. 또 Set을 만들고 map으로 바꾸는 과정도 말씀처럼 너무 우회적이라 가독성이 떨어지는 구조였던 것 같고요 😅

이번 기회에 말씀해주신 것처럼 VO끼리 직접 비교하는 방식으로 간결하게 개선해보겠습니다!
덕분에 VO의 사용 목적도 다시 한번 돌아보게 되었어요. 좋은 피드백 정말 감사합니다! 😊

Ohzzi commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,47 @@
+    private static final String ERROR_INVALID_SIZE = "[ERROR] 당첨 번호는 6개여야 합니다.";
+    private static final String ERROR_DUPLICATED = "[ERROR] 당첨 번호는 중복될 수 없습니다.";
+
+    private final List<LottoNumber> numbers;
+
```
만약 보너스 볼도 가진다면 WinningLotto가 단순 count만 반환하는게 아니라 Rank를 반환할 수 있지 않을까요?

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
말씀 듣고 보니, WinningLotto가 단순히 일치 개수만 반환하는 역할에 머무는 게 아니라, 보너스 번호까지 함께 가지고 있다면 최종 등수(Rank)를 판단하는 책임도 함께 가질 수 있겠다는 생각이 들었어요.

지금은 countMatch()에서 개수만 반환하고 있는데, Rank.valueOf(matchCount, bonusMatch)까지 연결되면 WinningLotto 쪽에서 바로 등수를 판단해주는 흐름으로 자연스럽게 확장될 수 있을 것 같아요!

이 구조가 더 도메인 중심적인 설계인 것 같아서, 이 방향으로 한 번 개선해보겠습니다. 좋은 인사이트 주셔서 정말 감사합니다 😊

Ohzzi commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,47 @@
+        return (int) count;
+    }
+
+    private void validate(List<LottoNumber> numbers) {
+        if (numbers.size() != REQUIRED_SIZE) {
+            throw new IllegalArgumentException(ERROR_INVALID_SIZE);
+        }
+
+        Set<LottoNumber> unique = new HashSet<>(numbers);
+        if (unique.size() != REQUIRED_SIZE) {
+            throw new IllegalArgumentException(ERROR_DUPLICATED);
+        }
+    }
```
여기가 사실 Lotto 클래스의 validation과 중복인 것 같은데요. WinningLotto가 Lotto를 가지면 중복을 해결할 수 있지 않을까요?

jsoonworld replied:
처음 이 코드를 작성할 때는 각 객체가 스스로 유효성을 판단하는 게 명확하겠다는 생각으로 중복을 감수하고 작성했었는데요!
말씀해주신 것처럼 WinningLotto가 Lotto를 가지도록 구조를 변경하면,
자연스럽게 중복을 제거할 수 있고 가시성이나 도메인 구조 측면에서도 더 직관적이겠다는 생각이 들었습니다 🙌

좋은 방향 제시해주셔서 감사합니다! 이 부분은 말씀해주신 대로 개선해보겠습니다 😊

Ohzzi commented on src/main/java/domain/Profit.java:
```diff
@@ -0,0 +1,44 @@
+
+public class Profit {
+    private static final String ERROR_ZERO_PURCHASE = "[ERROR] 구입 금액은 0보다 커야 합니다.";
+    private static final List<Rank> WINNING_RANKS = List.of(Rank.FIFTH, Rank.FOURTH, Rank.THIRD, Rank.SECOND, Rank.FIRST);
```
이 상수는 의미상 RANK에 public static으로 있는게 더 타당할 것 같아요. Profit과는 맞지 않는 것 같네요.

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
처음에는 수익 계산 로직에서만 사용하는 값이라고 판단해서 Profit 클래스 내부에 위치시켰는데요,
말씀해주신 것처럼 WINNING_RANKS는 단순히 로직용 상수가 아니라
"당첨으로 인정되는 등수"라는 도메인적인 의미를 담고 있어서, Rank 쪽에 위치시키는 게 훨씬 자연스럽고 응집력 있는 설계라는 생각이 들었습니다!

처음엔 values()로도 충분하지 않을까 생각했지만,
수익 계산처럼 당첨 등수만 반복하고 싶을 때는 WINNING_RANKS처럼 의미 있는 서브셋을 따로 정의해두면
의도도 명확해지고, 반복 구조도 더 깔끔해지는 장점이 있더라구요!

이번 경우처럼 서브셋의 도메인적 의미가 분명할 때는 상수화가 유효한 케이스라고 생각되어,
말씀해주신 방향대로 Rank 쪽으로 옮겨 적용해보겠습니다!
좋은 관점 공유해주셔서 감사합니다 😊

Ohzzi commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Lottos {
```
이 클래스가 하는 역할이 애매해보이네요. 그냥 수동 로또와 자동 로또를 하나의 리스트로 합쳤어도 충분하지 않았을까요? 일급 컬렉션이 가지는 의미와 역할에 대해 다시 한 번 생각해보면 좋을 것 같습니다.

jsoonworld replied:
처음에는 Lotto를 만들고, 복수 형태인 Lottos가 필요하겠다는 생각에 자연스럽게 일급 컬렉션을 만들어야겠다고 판단했는데요,
충분한 역할 정의 없이 배경 논리만으로 만들어지다 보니, 말씀해주신 대로 지금 와서 보면 다소 애매한 책임을 가지게 된 것 같습니다 😅

이번 기회에 일급 컬렉션이 가지는 의미와 역할에 대해서 다시 한 번 생각해보게 되었고, 간단히 아래와 같이 정리해보았습니다:

- 일급 컬렉션(First-class Collection)이란?
> "컬렉션을 감싸는 하나의 객체로, 컬렉션을 직접 노출하지 않고 의미 있는 행위와 제약을 그 안에서 정의하는 객체"

- 일급 컬렉션의 역할과 장점
1. 도메인 의미 부여
   - 단순한 List<Lotto> 대신 Lottos를 사용하면 "로또 목록"이라는 도메인 의미를 코드에 명확히 드러낼 수 있음
2. 불변성 보장
   - 컬렉션을 외부에 직접 노출하지 않고 필요한 연산만 메서드로 노출해 불변성을 유지
3. 로직 응집
   - 예: 로또 개수 계산, 로또 합치기(merge) 등 컬렉션 관련 로직을 한 곳에 응집 가능
   - List<Lotto>를 사용하는 곳마다 size()나 addAll()을 반복하는 대신, 의미 있는 count(), merge() 같은 메서드 제공 가능 
4. 중복 제거와 책임 분리
   - 여러 곳에서 동일하게 쓰이던 List<Lotto> 관련 로직을 이 클래스에 몰아넣어 중복 제거 가능

다만, 이번 경우처럼 일급 컬렉션을 만들긴 했지만 도메인에 기여하거나 책임을 갖지 못하고 단순히 리스트만 감싼 상태라면,
굳이 따로 만들기보다 그 자체를 합쳐서 쓰는 구조가 더 단순하고 명확할 수도 있다는 점도 배웠습니다.

말씀 덕분에 일급 컬렉션이 정말 필요한 상황은 어떤 케이스인지 다시 고민해보게 되었어요 🙏
꼼꼼한 피드백 정말 감사합니다! 이 부분은 좀 더 구조를 단순화하거나, 일급 컬렉션으로써 역할을 명확히 부여하는 쪽으로 개선해보겠습니다 😊

Ohzzi commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,47 @@
+        return value;
+    }
+
+    public static void validateBonusNumber(List<LottoNumber> winningNumbers, LottoNumber bonusNumber) {
+        for (LottoNumber number : winningNumbers) {
+            if (number.value() == bonusNumber.value()) {
+                throw new IllegalArgumentException(ERROR_DUPLICATE);
+            }
+        }
+    }
```
1. indent가 2네요!
2. 이 함수가 LottoNumber 클래스 안에 있어야 할 이유가 있을까요?
3. equals를 오버라이딩 했는데 value()를 호출하는 것은 불필요한 호출이 아닐까요?

jsoonworld replied:
아이코,, 인덴트 1로 수정하겠습니다! 😅

처음에는 당첨 번호와 보너스 번호 간의 중복을 단순히 정수값으로 비교하려고 value()를 꺼내 사용했는데요,
이미 LottoNumber에서 equals()를 오버라이딩하고 있기 때문에
객체 자체 비교만으로도 충분하다는 점을 간과했었네요.

또한 이 검증 로직이 LottoNumber 안에 위치하는 것도, 객체지향 관점에서는 어색한 선택이었던 것 같습니다.
보너스 번호가 당첨 번호와 중복되지 않아야 한다는 규칙은 LottoNumber의 역할이라기보단
WinningLotto가 책임져야 할 도메인 규칙이라는 점, 피드백 덕분에 다시 생각해볼 수 있었어요!

덕분에 역할과 책임을 다시 조정할 수 있었습니다. 🙏 감사합니다!

Ohzzi commented on src/main/java/service/LottoGenerator.java:
```diff
@@ -0,0 +1,38 @@
+
+    private Lottos generateAutoLottos(int autoCount) {
+        List<Lotto> tickets = IntStream.range(0, autoCount)
+                .mapToObj(i -> toLotto(numberGenerator.generate()))
```
numberGenerator.generate 하면 뭔가 숫자 하나가 나올 것 같아서 어색하네요

jsoonworld replied:
아이코 😅 의식적으로 의미 있게 작성하고 싶었는데, generate()라는 네이밍이 의도를 제대로 드러내지 못했던 것 같네요.

현재 메서드에서는 List<LottoNumber>를 생성하고 있는데, numberGenerator.generate()라고만 하면
정말 숫자 하나를 생성할 것처럼 보여서 반환값과 역할이 잘 매치되지 않는다는 점, 공감됩니당,,

실제로 아래처럼 사용할 때도:
```java
.mapToObj(i -> toLotto(numberGenerator.generate()))
```
generate()가 복수의 숫자 리스트를 반환한다는 걸 한눈에 파악하기가 어려워서,
generateLottoNumbers()처럼 더 구체적이고 의도를 잘 드러내는 이름이 훨씬 적절하다는 생각이 들었어요.

말씀해주신 덕분에 코드 읽는 사람 입장에서의 가독성도 다시 한번 돌아보게 되었습니다.
좋은 피드백 감사드리고, 의미가 명확하게 전달되도록 메서드명을 개선해보겠습니다! 🙇‍♂️

Ohzzi commented on src/test/java/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Test
+    @DisplayName("1부터 45 사이 값으로 생성할 수 있다")
+    void createWithValidValue() {
+        assertThatCode(() -> new LottoNumber(1)).doesNotThrowAnyException();
+        assertThatCode(() -> new LottoNumber(45)).doesNotThrowAnyException();
+        assertThat(new LottoNumber(10).value()).isEqualTo(10);
+    }
```
1하고 45 테스트 할때 value에 대한 테스트도 같이 하면 되지 않을까요?

jsoonworld replied:
흠... 말씀 듣고 보니 정말 그렇네요 😅
1과 45가 예외 없이 생성되는지만 확인하면 충분하다고 생각했는데,
value()까지 함께 검증해두면 생성뿐만 아니라 값이 정확히 설정되었는지도 확실히 알 수 있겠네요.

이런 부분은 미처 생각하지 못했는데 덕분에 테스트를 더 짜임새 있게 다듬을 수 있을 것 같아요!
앞으로는 경계값 테스트도 의도가 분명하게 드러나도록 더 신경 써서 작성해보겠습니다 🙇‍♂️
좋은 피드백 감사합니다!

Ohzzi commented on src/test/java/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,35 @@
+        assertThat(new LottoNumber(10).value()).isEqualTo(10);
+    }
+
+    @Test
+    @DisplayName("1 미만 값은 예외가 발생한다")
+    void lessThanMinThrowsException() {
+        assertThatThrownBy(() -> new LottoNumber(0))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("[ERROR] 번호는 1부터 45 사이여야 합니다.");
+    }
+
+    @Test
+    @DisplayName("45 초과 값은 예외가 발생한다")
+    void greaterThanMaxThrowsException() {
+        assertThatThrownBy(() -> new LottoNumber(46))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("[ERROR] 번호는 1부터 45 사이여야 합니다.");
+    }
```
사실 요런 케이스는 오히려 ParameterizedTest로 묶어도 되겠네요

jsoonworld replied:
오..! 말씀 듣고 보니 정말 그렇네요 🙇‍♂️
0이나 46처럼 예외가 발생해야 하는 케이스는 하나의 @ParameterizedTest로 묶는 편이 더 응집도 있게 표현할 수 있었겠네요!

이번에 바로 개선해보고, 다음 미션에서도 이런 부분까지 더 신경 써서 작성해보겠습니다 :)
피드백 감사합니다!

Ohzzi commented on src/test/java/domain/ProfitTest.java:
```diff
@@ -0,0 +1,104 @@
+        return new LottoStatisticsStub(countMap);
+    }
+
+    private static class LottoStatisticsStub extends LottoStatistics {
+        private final Map<Rank, Integer> countMap;
+
+        public LottoStatisticsStub(Map<Rank, Integer> countMap) {
+            super(
+                    new Lottos(List.of()),
+                    new WinningLotto(toLottoNumbers(List.of(1, 2, 3, 4, 5, 6))),
+                    new LottoNumber(7)
+            );
+            this.countMap = countMap;
+        }
+
+        @Override
+        public int countOf(Rank rank) {
+            return countMap.getOrDefault(rank, 0);
+        }
+    }
```
LottoStatistics에 Map을 받는 생성자를 여는게 더 편하지 않았을까요?

jsoonworld replied:
프러덕션 코드에서는 필요하지 않다고 판단해서 테스트용 생성자를 두지 않고 LottoStatisticsStub 클래스를 만들어 사용했었는데요, 돌이켜보면 테스트도 관리 대상이라는 점에서, 테스트 목적에 맞는 생성자를 열어두는 게 작성자나 리뷰어 입장에서 모두 더 좋았을 것 같아요 😊

LottoStatisticsStub처럼 클래스를 따로 만들다 보니 코드가 길어지고,
정작 테스트 목적에 비해 과한 구조가 되었던 점도 있었네요.

말씀 주신 것처럼 Map<Rank, Integer>를 받는 생성자를 LottoStatistics에 열어두면
테스트에서 new LottoStatistics(Map.of(...))처럼 간결하게 사용할 수 있어서
테스트의 명확성과 유지보수 측면에서도 훨씬 나은 방향이라고 느꼈습니다!

이번 피드백 덕분에 테스트 편의성과 실제 도메인 설계 사이의 균형도 다시 한번 생각해볼 수 있었어요. 감사합니다 🙇‍♂️

Ohzzi commented on src/test/java/domain/RankTest.java:
```diff
@@ -0,0 +1,49 @@
+    @Test
+    @DisplayName("일치 개수에 따라 보너스 없이 적절한 Rank를 반환한다")
+    void matchCountOf_returnsCorrectRank() {
+        SoftAssertions softly = new SoftAssertions();
```
오 이런게 있는줄은 저도 몰랐네요!! 👍

jsoonworld replied:
하하~ 저도 이번에 처음 써봤습니다 😄

이전에 
> "검증문이 굉장히 많은 상황입니다. 검증문이 많다면 soft assertion을 사용하는 편이 좀 더 유리하지 않을까요?" 

라고 조언 주신 게 기억나서, 이번 기회에 SoftAssertions를 찾아보고 실제로 적용해보게 되었어요!

확실히 여러 값을 한꺼번에 검증해야 할 때,
실패 시 어떤 값이 잘못됐는지 한눈에 파악할 수 있어서 테스트 품질 면에서도 도움이 많이 되더라고요.
앞으로 자주 활용하게 될 것 같습니다!  🙇‍♂️

Ohzzi commented on src/test/java/domain/RankTest.java:
```diff
@@ -0,0 +1,49 @@
+        assertThat(Rank.valueOf(count, false)).isEqualTo(Rank.MISS);
+    }
+
+    @Test
+    @DisplayName("당첨 Rank는 isWinning()이 true를 반환한다")
+    void isWinning_returnsTrueForWinningRanks() {
+        SoftAssertions softly = new SoftAssertions();
+        softly.assertThat(Rank.FIFTH.isWinning()).isTrue();
+        softly.assertThat(Rank.FOURTH.isWinning()).isTrue();
+        softly.assertThat(Rank.THIRD.isWinning()).isTrue();
+        softly.assertThat(Rank.SECOND.isWinning()).isTrue();
+        softly.assertThat(Rank.FIRST.isWinning()).isTrue();
+        softly.assertAll();
+    }
```
이경우는 EnumSource를 쓰는게 더 깔끔할 것 같습니다!

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
이렇게 반복적인 enum 값에 대한 검증은 EnumSource를 활용하면 훨씬 깔끔하게 표현할 수 있겠네요!

이번에는 SoftAssertions로 여러 값을 한 번에 검증해보자는 생각이 먼저 들어 적용했는데,
말씀해주신 방식이 더 간결하고 확장성도 좋을 것 같다는 생각이 들었습니다.

테스트 코드도 읽기 쉬운 방향으로 계속 개선해보겠습니다!  😊

Ohzzi commented on src/test/java/service/LottoGeneratorTest.java:
```diff
@@ -0,0 +1,86 @@
+        assertThat(result.count()).isEqualTo(manualLottos.size() + autoCount);
+    }
+
+    @Test
+    @DisplayName("자동 로또만 생성되어야 한다 (수동 로또 없이)")
+    void generate_withAutoLottosOnly_shouldReturnCorrectAutoLottos() {
+        // given
+        List<Lotto> manualLottos = List.of();
+        int autoCount = 3;
+
+        // when
+        Lottos result = lottoGenerator.generate(manualLottos, autoCount);
+
+        // then
+        assertThat(result.count()).isEqualTo(autoCount);
+    }
+
+    @Test
+    @DisplayName("자동 로또만 생성되어야 한다 (수동 로또 없이, 개수 5개)")
+    void generate_withEmptyManualLottos_shouldReturnOnlyAutoLottos() {
+        // given
+        List<Lotto> manualLottos = List.of();
+        int autoCount = 5;
+
+        // when
+        Lottos result = lottoGenerator.generate(manualLottos, autoCount);
+
+        // then
+        assertThat(result.count()).isEqualTo(autoCount);
+    }
```
두 테스트는 무엇이 다른가요?

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
두 테스트가 모두 "수동 로또 없이 자동 로또만 생성하는 상황"이라는 점에서 본질적으로 같은 시나리오를 다루고 있었네요.

처음엔 개수를 달리한 테스트로 각각의 케이스를 더 명확히 표현하고 싶었는데, 말씀처럼 단순히 개수만 다른 반복 테스트라면 하나의 테스트로 충분히 커버 가능했을 것 같습니다.

특히 이 경우엔 테스트 이름도 중복되고 테스트 목적도 완전히 겹치기 때문에,
두 테스트를 통합하고 @ParameterizedTest로 리팩토링하는 방향이 훨씬 깔끔하고 유지보수에도 유리하겠네요!

앞으로는 테스트 중복 여부를 더 잘 살펴보고, 의도가 겹친다면 하나의 테스트로 응집력 있게 구성해보겠습니다 😊

Ohzzi commented on src/test/java/service/LottoServiceTest.java:
```diff
@@ -0,0 +1,67 @@
+    @Test
+    @DisplayName("구매 금액 10000원, 수동 로또 2개를 입력받아 로또 구매 요청을 제대로 생성한다")
+    void createLottoPurchaseRequest_ShouldPreparePurchaseDto() {
+        // Given
+        int totalAmount = 10000;
+        int manualLottoCount = 2;
+        int autoLottoCount = 8;
+
+        // When
+        LottoPurchaseDto purchaseRequest = lottoService.preparePurchase(inputView, outputView);
+
+        // Then
+        assertEquals(totalAmount, purchaseRequest.totalAmount());
+        assertEquals(manualLottoCount, purchaseRequest.manualLottoCount());
+        assertEquals(autoLottoCount, purchaseRequest.autoLottoCount());
```
given과 then이 뭔가 좀 이상한 것 같네요.

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
처음에는 테스트 값들을 미리 변수로 선언하고 Given 블록처럼 분리하는 방식이 더 명시적이라고 생각했는데,
말씀해주신 것처럼 오히려 when에서 생성한 결과값에 직접 값을 넣어 검증하는 게 흐름상 더 깔끔하고 자연스럽네요!

이번에는 테스트 대상의 입력을 StubInputView에서 이미 고정하고 있는 구조이기도 해서,
별도로 totalAmount, manualLottoCount 같은 값을 선언하지 않고도 의미 전달이 충분하겠다는 생각이 들었습니다.
테스트 구조도 더 간결해지고, 의도도 명확히 드러날 수 있도록 개선해보았습니다.

꼼꼼스 피드백 감사합니다! 다음 미션에서도 이런 흐름을 좀 더 의식하며 작성해보겠습니다 😊

Ohzzi commented on src/test/java/utils/WinningLottoParserTest.java:
```diff
@@ -0,0 +1,65 @@
+        WinningLotto result = WinningLottoParser.parse("1, 2, 3, 4, 5, 6");
+
+        assertThat(result.countMatch(toLotto(List.of(1, 2, 3, 10, 11, 12))))
+                .isEqualTo(3);
```
테스트에 맞지 않는 검증문이 아닐까요?

jsoonworld replied:
피드백 감사합니다! 🙇‍♂️
말씀해주신 것처럼 이 테스트는 WinningLottoParser가 올바르게 파싱되는지 확인하는 게 목적이라,
countMatch()를 호출해서 일치 개수를 검증하는 부분은 테스트 목적과 조금 동떨어진 내용이었던 것 같아요.

의도는 생성된 WinningLotto 객체의 정상적인 동작까지 간단히 확인하고 싶어서 추가했던 검증이었는데,
테스트의 책임을 벗어나는 로직까지 포함되면서 오히려 혼란을 줄 수 있겠다는 생각이 들었습니다.

해당 부분은 제거하고, 파싱 로직 자체의 책임에 집중한 테스트로 정리해보겠습니다!

Ohzzi reviewed (changes requested):
대대적인 리팩터링 하느라 수고가 많으셨습니다 장순님! 👍 그에 따라 리뷰를 남겨봤는데요, 세세한 리뷰는 각 코멘트에 남겨져 있습니다. 다만 전체적으로 피드백을 남기자면,

**여전히 코드가 과합니다.** 미션에 주어지는 요구사항의 볼륨과 난이도 대비 장순님의 코드는 한 번의 리팩토링을 거친 이후로도 제 눈에는 오버 엔지니어링으로 보여요. 기존 대비 클래스 수를 1/3 가까이 줄여주셨음에도 불구하고요. 정말 필요한 클래스만 만든다면 지금보다도 훨씬 많이 줄일 수 있을거에요. 단순 클래스 개수만을 말씀드리는 것이 아니라, 메서드 분리도, 테스트 케이스도 마찬가지입니다.

물론 이런 생각이 들 수 밖에 없습니다.

- 확장성을 챙겨야 하니까 최대한 추상화를 해야지
- 테스트 케이스가 불안한데? 중간값들도 테스트를 몇 개 넣어야겠다.
- 단일 책임 원칙에 의하면 클래스는 하나의 일만 해야 하는데, 이거 지금 두 가지 일을 하는 것이 아닐까? 클래스를 분리해야겠다.

이것은 분명 오답은 아닙니다만, 제 개인적으로는 좋은 코드 습관이라고는 생각하지 않습니다. 추상화를 하고, 테스트를 무수히 많이 만들고, 클래스를 더 잘게 쪼개고. 그런 것들이 단점 없이 장점만 있다면 모두가 그렇게 했겠지만 그렇지 않습니다. 결국 코드는 잘게 쪼개면 쪼갤수록 한 눈에 파악하기 어렵고 복잡해지는 결과를 낳거든요. 분리로 인해 얻는 이득이 손실보다 크지 않다면, 그것을 적절하다고 보기는 어려울 것 같습니다. 오버 엔지니어링이라고 볼 수 있을 것 같아요.

오버 엔지니어링을 경계해야 하는 이유로는 하나는 협업의 어려움을 꼽을 수 있고, 하나는 실무라는 것은 정해진 기한이 있기 때문을 꼽을 수 있습니다. 여러 사람이 한 코드를 함께 개발해야 하는 협업의 특성 상, 내가 오버 엔지니어링 하는 만큼 다른 사람이 그 코드를 파악하는데 곱절의 노력을 들여야 합니다. 또한 실무로 가면 빡빡한 기한 내에 요구사항을 만족시켜야 하는데, 오버 엔지니어링이 들어가면 산출물을 내놓는데 걸리는 시간이 그만큼 늘어나게 됩니다.

때문에 좋은 코드에는 이유와 근거가 있어야 한다고 생각합니다. 그런데 현재 장순님의 코드는 대체적으로 `분리의 근거`가 부족한 것 같아요. 물론 의식적으로 메서드, 클래스, 인터페이스를 분리해보는 연습도 필요하다고 생각하지만, 제가 생각했을 때 장순님의 자바 개발 실력은 의식적인 연습이 필요한 레벨보다는 더 높이 있다고 생각하기 때문에 장순님께는 필요한 것 같지는 않습니다. 그래서 다음 미션에서는 **기능이 돌아가도록 우선은 최대한 작게 만들어보고 리팩토링하는 과정에서 조금씩 `정말 필요한 분리`를 해나가면서 진행해보는 것**이 좋을 것 같아요.

머지는 2차 리뷰사항들 확인하시면 요청드리도록 하겠습니다! 다음 미션도 파이팅입니다 💪

jsoonworld committed:
refactor(feedback) : 2차 코드 리뷰 피드백 반영

Ohzzi reviewed (approved):
👍 👍 수고 많으셨어요!! 리뷰 사항들 다음 미션에서는 잘 체득한 상태로 만나도록 합시다!!
